### PR TITLE
Add Microsoft Azure Enumeration Auxiliary Module

### PR DESCRIPTION
**Pull request by RoseSecurity**
This pull request adds an auxiliary module for enumerating public Azure services by validating legitimate subdomains through DNS record queries. Microsoft makes use of a number of different domains and subdomains for each of their Azure services. This module identifies API services, storage accounts, key vaults, and databases.

## Verification

```
msf6 > use auxiliary/gather/enum_azuresubdomains
msf6 auxiliary(gather/enum_azuresubdomains) > show options 

Module options (auxiliary/gather/enum_azuresubdomains):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   DOMAIN                         yes       The target domain without TLD (Ex: victim rather than victim.org)
   ENUM_A        true             yes       Enumerate DNS A record
   ENUM_CNAME    true             yes       Enumerate DNS CNAME record
   ENUM_MX       true             yes       Enumerate DNS MX record
   ENUM_NS       true             yes       Enumerate DNS NS record
   ENUM_SOA      true             yes       Enumerate DNS SOA record
   ENUM_TXT      true             yes       Enumerate DNS TXT record
   NS                             no        Specify the nameservers to use for queries, space separated
   PERMUTATIONS  false            no        Prepend and append permutated keywords to domain (This option can take minutes to complete)
   Proxies                        no        A proxy chain of format type:host:port[,type:host:port][...]
   RPORT         53               yes       The target port (TCP)
   SEARCHLIST                     no        DNS domain search list, comma separated
   THREADS       1                yes       Number of threads to use in threaded queries


View the full module info with the info, or info -d command.

msf6 auxiliary(gather/enum_azuresubdomains) > set DOMAIN umuc365
DOMAIN => umuc365
msf6 auxiliary(gather/enum_azuresubdomains) > set PERMUTATIONS true 
PERMUTATIONS => true
msf6 auxiliary(gather/enum_azuresubdomains) > run

[*] Discovered Target Domain: umuc365.mail.protection.outlook.com 

[*] Querying DNS CNAME records for umuc365.mail.protection.outlook.com
[*] Querying DNS NS records for umuc365.mail.protection.outlook.com
[*] Querying DNS MX records for umuc365.mail.protection.outlook.com
[*] Querying DNS SOA records for umuc365.mail.protection.outlook.com
[*] Querying DNS TXT records for umuc365.mail.protection.outlook.com

[*] Discovered Target Domain: umuc365.sharepoint.com 

[*] Querying DNS CNAME records for umuc365.sharepoint.com
[+] umuc365.sharepoint.com CNAME: 2732-ipv4v6e.clump.dprodmgd105.aa-rt.sharepoint.com
[*] Querying DNS NS records for umuc365.sharepoint.com
[*] Querying DNS MX records for umuc365.sharepoint.com
[*] Querying DNS SOA records for umuc365.sharepoint.com
[*] Querying DNS TXT records for umuc365.sharepoint.com

[*] Discovered Target Domain: umuc365-web.sharepoint.com 

[*] Querying DNS CNAME records for umuc365-web.sharepoint.com
[+] umuc365-web.sharepoint.com CNAME: umuc365.sharepoint.com
[*] Querying DNS NS records for umuc365-web.sharepoint.com
[*] Querying DNS MX records for umuc365-web.sharepoint.com
[*] Querying DNS SOA records for umuc365-web.sharepoint.com
[*] Querying DNS TXT records for umuc365-web.sharepoint.com
[*] Auxiliary module execution completed
```

### CODE REVIEW

File: modules/auxiliary/gather/enum_azuresubdomains.rb [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
```
Comment (jheysel-r7) [inline modules/auxiliary/gather/enum_azuresubdomains.rb:1]: Is this still necessary?

**Review by jheysel-r7 — COMMENTED** [review]
Hey @RoseSecurity, thanks so much for the contribution! We really appreciate it.

I've tested the module, seems to be working nicely. A couple things required before we can land this - as mentioned by the bots above, a documentation file and some linting. Let me know if you have any questions about either of those I'd be happy to assist.

```
msf6 auxiliary(gather/enum_azuresubdomains) > set domain testing12345678
domain => testing12345678
msf6 auxiliary(gather/enum_azuresubdomains) > run

[+] Discovered Target Domain: testing12345678.scm.azurewebsites.net

[*] Querying DNS CNAME records for testing12345678.scm.azurewebsites.net
[+] testing12345678.scm.azurewebsites.net CNAME: waws-prod-dm1-189.sip.azurewebsites.windows.net
[*] Querying DNS NS records for testing12345678.scm.azurewebsites.net
[*] Querying DNS MX records for testing12345678.scm.azurewebsites.net
[*] Querying DNS SOA records for testing12345678.scm.azurewebsites.net
[*] Querying DNS TXT records for testing12345678.scm.azurewebsites.net
[+] Discovered Target Domain: testing12345678.azurewebsites.net

[*] Querying DNS CNAME records for testing12345678.azurewebsites.net
[+] testing12345678.azurewebsites.net CNAME: hosts.testing12345678.azurewebsites.net
[*] Querying DNS NS records for testing12345678.azurewebsites.net
[*] Querying DNS MX records for testing12345678.azurewebsites.net
[*] Querying DNS SOA records for testing12345678.azurewebsites.net
[*] Querying DNS TXT records for testing12345678.azurewebsites.net
[+] Discovered Target Domain: testing12345678.file.core.windows.net

[*] Querying DNS CNAME records for testing12345678.file.core.windows.net
[+] testing12345678.file.core.windows.net CNAME: file.blz22prdstr06a.store.core.windows.net
[*] Querying DNS NS records for testing12345678.file.core.windows.net
[*] Querying DNS MX records for testing12345678.file.core.windows.net
[*] Querying DNS SOA records for testing12345678.file.core.windows.net
[*] Querying DNS TXT records for testing12345678.file.core.windows.net
[+] Discovered Target Domain: testing12345678.blob.core.windows.net

[*] Querying DNS CNAME records for testing12345678.blob.core.windows.net
[+] testing12345678.blob.core.windows.net CNAME: blob.blz22prdstr06a.store.core.windows.net
[*] Querying DNS NS records for testing12345678.blob.core.windows.net
[*] Querying DNS MX records for testing12345678.blob.core.windows.net
[*] Querying DNS SOA records for testing12345678.blob.core.windows.net
[*] Querying DNS TXT records for testing12345678.blob.core.windows.net
[+] Discovered Target Domain: testing12345678.queue.core.windows.net

[*] Querying DNS CNAME records for testing12345678.queue.core.windows.net
[+] testing12345678.queue.core.windows.net CNAME: queue.blz22prdstr06a.store.core.windows.net
[*] Querying DNS NS records for testing12345678.queue.core.windows.net
[*] Querying DNS MX records for testing12345678.queue.core.windows.net
[*] Querying DNS SOA records for testing12345678.queue.core.windows.net
[*] Querying DNS TXT records for testing12345678.queue.core.windows.net
[+] Discovered Target Domain: testing12345678.table.core.windows.net

[*] Querying DNS CNAME records for testing12345678.table.core.windows.net
[+] testing12345678.table.core.windows.net CNAME: table.blz22prdstr06a.store.core.windows.net
[*] Querying DNS NS records for testing12345678.table.core.windows.net
[*] Querying DNS MX records for testing12345678.table.core.windows.net
[*] Querying DNS SOA records for testing12345678.table.core.windows.net
[*] Querying DNS TXT records for testing12345678.table.core.windows.net
[+] Discovered Target Domain: testing12345678.mail.protection.outlook.com

[*] Querying DNS CNAME records for testing12345678.mail.protection.outlook.com
[*] Querying DNS NS records for testing12345678.mail.protection.outlook.com
[*] Querying DNS MX records for testing12345678.mail.protection.outlook.com
[*] Querying DNS SOA records for testing12345678.mail.protection.outlook.com
[*] Querying DNS TXT records for testing12345678.mail.protection.outlook.com
[+] Discovered Target Domain: testing12345678.database.windows.net

[*] Querying DNS CNAME records for testing12345678.database.windows.net
[+] testing12345678.database.windows.net CNAME: dataslice2.ukwest.database.windows.net
[*] Querying DNS NS records for testing12345678.database.windows.net
[*] Querying DNS MX records for testing12345678.database.windows.net
[*] Querying DNS SOA records for testing12345678.database.windows.net
[*] Querying DNS TXT records for testing12345678.database.windows.net
[*] Auxiliary module execution completed
```

**Comment by RoseSecurity** [comment]
Hey @jheysel-r7! Thanks for assisting with this module as it is my first one! I just added the required documentation and performed linting with Rubocop and MSFTidy. For Rubocop, it stated that I need to add a *Notes* section under the module's information for Stability and Reliability, but since my module only performs enumeration and is not an exploit, is this still required?

**Comment by adfoster-r7** [comment]
> I need to add a Notes section under the module's information for Stability and Reliability, but since my module only performs enumeration and is not an exploit, is this still required?

For now we've been using the pattern of providing empty values for the notes section

```
        'Notes' => {
          'Stability' => [],
          'SideEffects' => [],
          'Reliability' => []
        }
```

**Comment by jmartin-r7** [comment]
This seems like it could fit as enhancements to support a `permutaions` option in [auxiliary/gather/enum_dns](https://github.com/rapid7/metasploit-framework/blob/2de3142cdefb08a5670d27216e2afb26e47c6529/modules/auxiliary/gather/enum_dns.rb) and have the lists here extracted as wordlists files that can simply be set a options.

**Comment by jheysel-r7** [comment]
Hey @RoseSecurity, after discussing with the team we've come to the conclusion that this addition would likely be better off as an enhancement to the existing [auxiliary/gather/enum_dns](https://github.com/rapid7/metasploit-framework/blob/2de3142cdefb08a5670d27216e2afb26e47c6529/modules/auxiliary/gather/enum_dns.rb) module. You'll notice they're similar although `enum_dns` currently doesn't support `permutations` as of now. 

The list of public Azure services could be added to `metasploit-framework/data/wordlists` and then set  in the module options. Let me know if you have any questions regarding this!

**Comment by RoseSecurity** [comment]
Hey @jheysel-r7, I get that, is there anyway I can contribute moving forward?

**Comment by jmartin-r7** [comment]
@RoseSecurity you can refactor this PR based on the requested pattern. This idea in this PR looks like a nice enhancement in capabilities.

**Comment by jheysel-r7** [comment]
Hey @RoseSecurity, just checking in. Wondering how the if you had any question regarding the refactoring of the[auxiliary/gather/enum_dns](https://github.com/rapid7/metasploit-framework/blob/2de3142cdefb08a5670d27216e2afb26e47c6529/modules/auxiliary/gather/enum_dns.rb) module?

**Comment by RoseSecurity** [comment]
Hey @jheysel-r7, I have been a little busy with work, but just to clarify, you would like a `PERMUTATIONS_FILE` option added  to the [auxiliary/gather/enum_dns](https://github.com/rapid7/metasploit-framework/blob/2de3142cdefb08a5670d27216e2afb26e47c6529/modules/auxiliary/gather/enum_dns.rb) with wordlists for the permutations?

**Comment by jheysel-r7** [comment]
> Hey @jheysel-r7, I have been a little busy with work, but just to clarify, you would like a `PERMUTATIONS_FILE` option added to the [auxiliary/gather/enum_dns](https://github.com/rapid7/metasploit-framework/blob/2de3142cdefb08a5670d27216e2afb26e47c6529/modules/auxiliary/gather/enum_dns.rb) with wordlists for the permutations?

@RoseSecurity no worries, we appreciate you taking the time to do this. Yes that's correct. Please merge the functionality of this PR into [auxiliary/gather/enum_dns](https://github.com/rapid7/metasploit-framework/blob/2de3142cdefb08a5670d27216e2afb26e47c6529/modules/auxiliary/gather/enum_dns.rb) via a module option. The lists stored in arrays currently can be extracted as wordlists files that can be set as module options as well. Thanks!

**Pull request closed without merging**